WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Package Details/ShippingLabelPackageDetailsViewModel.swift

```diff
@@ -1,50 +1,148 @@
+    ///
+    private var productVariations: [ProductVariation] = []
+
+    /// The weight unit used in the Store
+    ///
+    let weightUnit: String?
+
+    /// The items rows observed by the main view `ShippingLabelPackageDetails`
+    ///
+    @Published var itemsRows: [ItemToFulfillRow] = []
+
+    init(order: Order,
+         formatter: CurrencyFormatter = CurrencyFormatter(currencySettings: ServiceLocator.currencySettings),
+         stores: StoresManager = ServiceLocator.stores,
+         storageManager: StorageManagerType = ServiceLocator.storageManager,
+         weightUnit: String? = ServiceLocator.shippingSettingsService.weightUnit) {
+        self.order = order
+        self.orderItems = order.items
+        self.currency = order.currency
+        self.currencyFormatter = formatter
+        self.stores = stores
+        self.storageManager = storageManager
+        self.weightUnit = weightUnit
+        configureResultsControllers()
+        syncProducts()
+        syncProductVariations()
+    }
+
+    private func configureResultsControllers() {
+        resultsControllers = ShippingLabelPackageDetailsResultsControllers(siteID: order.siteID,
+                                                                           orderItems: order.items,
+                                                                           storageManager: storageManager,
+           onProductReload: { [weak self] (products) in
+            guard let self = self else { return }
+            self.products = products
+            self.itemsRows = self.generateItemsRows()
+        }, onProductVariationsReload: { [weak self] (productVariations) in
+            guard let self = self else { return }
+            self.productVariations = productVariations
+            self.itemsRows = self.generateItemsRows()
+        })
+
+        products = resultsControllers?.products ?? []
+        productVariations = resultsControllers?.productVariations ?? []
+        itemsRows = generateItemsRows()
+    }
+
+    /// Generate the items rows, creating an element in the array for every item (eg. if there is an item with quantity 3,
+    /// we will generate 3 different items), and we will remove virtual products. 
+    ///
+    private func generateItemsRows() -> [ItemToFulfillRow] {
+        var itemsToFulfill: [ItemToFulfillRow] = []
+        for item in orderItems {
+            let isVariation = item.variationID > 0
+            var product: Product?
+            var productVariation: ProductVariation?
+
+            if isVariation {
+                productVariation = productVariations.first { $0.productVariationID == item.variationID }
+            }
+            else {
+                product = products.first { $0.productID == item.productID }
+            }
+            if product?.virtual == false || productVariation?.virtual == false {
+                for _ in 0..<item.quantity.intValue {
```

Comment:
If my order contains a fractional quantity of an item (e.g. 3.5) and we get a row for each whole item (3), should we not also get a row for the fractional remainder (0.5)?

Also, maybe explain in a comment why we are row-izing items with quantity != 1?

Reply:
Good question. I double-checked this on Core, and it seems that the quantity is always an int value. Also if you create an Order from the core (so on the admin side) and you add a fractional quantity, it will be converted to an int value. So the real question is why `quantity` is considered a Decimal value. I opened an issue about it https://github.com/woocommerce/woocommerce-ios/issues/3908

> Also, maybe explain in a comment why we are row-izing items with quantity != 1?

Done ✅

Reply:
Discussion about this, here https://a8c.slack.com/archives/C6H8C3G23/p1617703217370100

Reply:
So, after some investigations, it seems that I handled this particular case in the right way. The web client considers a quantity like `3.5` as `3 items`. 

<img width="1124" alt="Screen Shot 2021-04-06 at 12 12 47" src="https://user-images.githubusercontent.com/495617/113708858-0f653b80-96e2-11eb-8e22-76b5a1c8c682.png">